packages/curve-ui-kit/src/features/connect-wallet/lib/ConnectionContext.tsx

```diff
@@ -0,0 +1,178 @@
+  FAILURE: 'failure',
+}
+
+export const CONNECT_STAGE = {
+  CONNECT_API: 'api',
+  HYDRATE: 'hydrate',
+  CONNECT_WALLET: 'connect-wallet',
+  DISCONNECT_WALLET: 'disconnect-wallet',
+  SWITCH_NETWORK: 'switch-network',
+} as const
+
+export type ConnectStage = (typeof CONNECT_STAGE)[keyof typeof CONNECT_STAGE]
+export type ConnectStatus = (typeof CONNECT_STATUS)[keyof typeof CONNECT_STATUS]
+
+export type ConnectState = {
+  status: ConnectStatus
+  stage?: ConnectStage
+}
+
+const { FAILURE, LOADING, SUCCESS } = CONNECT_STATUS
+const { HYDRATE, SWITCH_NETWORK, CONNECT_WALLET, CONNECT_API } = CONNECT_STAGE
+
+export const isSuccess = (connectState: ConnectState) => connectState.status === SUCCESS
+
+export const isFailure = ({ status, stage: connectionStage }: ConnectState, expectedStage?: string) =>
+  status === FAILURE && Boolean(!expectedStage || connectionStage?.startsWith(expectedStage))
+
+export const isLoading = ({ status, stage: connectionStage }: ConnectState, expectedStage?: string | string[]) =>
+  status === LOADING &&
+  Boolean(
+    !expectedStage ||
+      (Array.isArray(expectedStage)
+        ? expectedStage.some((s) => connectionStage?.startsWith(s))
+        : connectionStage?.startsWith(expectedStage)),
+  )
+
+type ConnectionContextValue<TLib> = {
+  connectState: ConnectState
+  lib: TLib | null
+}
+
+const ConnectionContext = createContext<ConnectionContextValue<unknown>>({
+  connectState: { status: LOADING },
+  lib: null,
+})
+
+/**
+ * ConnectionProvider is a React context provider that manages the connection state of a wallet.
+ * We use a context instead of a store to be able to get the initialization functions injected depending on the app.
+ * todo: Merged with useWallet after wagmi migration. Get rid of apiStore after this is used everywhere.
+ */
+export const ConnectionProvider = <
+  TChainId extends number,
+  TLib extends { chainId: TChainId; signerAddress?: string },
+>({
+  hydrate,
+  initLib,
+  chainId,
+  onChainUnavailable,
+  children,
+}: {
+  hydrate: (newLib: TLib, prevLib: TLib | null, wallet: Wallet | null) => Promise<void>
+  initLib: (chainId: TChainId, wallet: Wallet | null) => Promise<TLib>
+  chainId: TChainId
+  onChainUnavailable: ([unsupportedChainId, walletChainId]: [TChainId, TChainId]) => void
+  children: ReactNode
+}) => {
+  const [connectState, setConnectState] = useState<ConnectState>({ status: LOADING })
+  const { wallet, connect } = useWallet()
+  const [_, setChain] = useSetChain()
+
+  const walletRef = useRef(wallet) // use ref to avoid re-rendering on wallet change
+  walletRef.current = wallet
+
+  const walletChainId = getWalletChainId(wallet)
+  const walletSignerAddress = getWalletSignerAddress(wallet)
+  const walletLabel = wallet?.label
+
+  useEffect(() => {
+    const abortController = new AbortController()
+    const walletName = getFromLocalStorage<string>(WalletNameStorageKey) // todo: we might not need the walletName at all in useWallet
```

Comment:
I believe we can remove wallet names and such the moment we get rid of onboard. Wagmi at least doesn't need it.

Reply:
yes and the reconnect logic totally!